// gennum-peodb/lib/getIDs.js
//============================================================================//

var print = console.log;

//============================================================================//

module.exports = function(PEODB) {

  function getIdList(db, tableName, idName, cb) {
    var qStr = 'SELECT `' + idName + '`'
              + ' FROM `' + tableName + '`';
    return db.queryData(qStr, function(err, data) {
      if (err) {
        return cb(err);
      }
      return cb(null, data.toArray());
    });
  }

  PEODB.prototype.getUserIDs = function(cb) {
    return getIdList(this.sharedDB, 'users', 'user_id', cb);
  }

  PEODB.prototype.getPartflowIDs = function(cb) {
    return getIdList(this.sharedDB, 'partflows', 'partflow_id', cb);
  }

  PEODB.prototype.getPartTypeIDs = function(cb) {
    return getIdList(this.sharedDB, 'part_types', 'part_type_id', cb);
  }

  PEODB.prototype.getPartIDs = function(cb) {
    return getIdList(this.sharedDB, 'parts', 'part_id', cb);
  }

  PEODB.prototype.getStepTypeIDs = function(cb) {
    return getIdList(this.sharedDB, 'step_types', 'step_type_id', cb);
  }

  PEODB.prototype.getStepIDs = function(cb) {
    return getIdList(this.sharedDB, 'steps', 'step_id', cb);
  }

  PEODB.prototype.getPartOutputIDs = function(cb) {
    return getIdList(this.sharedDB, 'part_outputs', 'part_output_id', cb);
  }

  PEODB.prototype.getStepOutputIDs = function(cb) {
    return getIdList(this.sharedDB, 'step_outputs', 'step_output_id', cb);
  }

  PEODB.prototype.getTPsetIDs = function(cb) {
    return getIdList(this.sharedDB, 'tp_sets', 'tp_set_id', cb);
  }

  PEODB.prototype.getTPIDs = function(cb) {
    return getIdList(this.sharedDB, 'tps', 'tp_id', cb);
  }

  PEODB.prototype.getTPrevIDs = function(cb) {
    return getIdList(this.sharedDB, 'tp_revs', 'tp_rev_id', cb);
  }
}

//============================================================================//
