// gennum-peodb/lib/getObjectID.js

var peodbUtils = require('../peodbUtils');

//============================================================================//

function getID(args, name) {
  var val = args[name];
  if (typeof(val) !== 'undefined') {
    return val;
  }
  return args;
}

//============================================================================//

module.exports = function(PEODB) {

  PEODB.prototype.getObjectID = function(args, type, cb) {
    if (typeof(type) === 'function') {
      type = args.type;
      cb   = type;
    }
    var id = getID(args, type + '_id');
    var db = args.db || this.sharedDB;
    //console.log('PEODB.prototype.getObjectID', type, id, args); // FIXME
    return peodbUtils.getObjectID(db, type, id, cb);
  }

  //==========================================================================//

  PEODB.prototype.getPartflowObjectID = function(args, cb) {
    var id = getID(args, 'partflow_id');
    var db = args.db || this.sharedDB;
    return peodbUtils.getPartflowObjectID(db, id, cb);
  }

  PEODB.prototype.getPartObjectID = function(args, cb) {
    var id = getID(args, 'part_id');
    var db = args.db || this.sharedDB;
    return peodbUtils.getPartObjectID(db, id, cb);
  }

  PEODB.prototype.getStepObjectID = function(args, cb) {
    var id = getID(args, 'step_id');
    var db = args.db || this.sharedDB;
    return peodbUtils.getStepObjectID(db, id, cb);
  }

  PEODB.prototype.getTPsetObjectID = function(args, cb) {
    var id = getID(args, 'tp_set_id');
    var db = args.db || this.sharedDB;
    return peodbUtils.getTPsetObjectID(db, id, cb);
  }

  PEODB.prototype.getTPObjectID = function(args, cb) {
    var id = getID(args, 'tp_id');
    var db = args.db || this.sharedDB;
    return peodbUtils.getTPObjectID(db, id, cb);
  }

  PEODB.prototype.getTPrevObjectID = function(args, cb) {
    var id = getID(args, 'tp_rev_id');
    var db = args.db || this.sharedDB;
    return peodbUtils.getTPrevObjectID(db, id, cb);
  }

  PEODB.prototype.getFabProcessObjectID = function(args, cb) {
    var id = getID(args, 'fab_process_id');
    var db = args.db || this.sharedDB;
    return peodbUtils.getFabProcessObjectID(db, id, cb);
  }

  //==========================================================================//
}

//============================================================================//
